scripts: nrf_profiler: Document stats calculating script (#24424)
nrfconnect/sdk-nrf: scripts/nrf_profiler/README.rst

   An application must provide its own implementation.
   See the :ref:`nrf_desktop_nrf_profiler_sync` for an example implementation used by the :ref:`nrf_desktop` application.

.. _nrf_profiler_script_calculating_statistics:

Calculating statistics
======================

The :file:`calc_stats.py` script is used to calculate and visualize the nRF Profiler event propagation statistics (statistics for time intervals between nRF Profiler events) for a given dataset.
A JSON file containing test presets specifies the events to be analyzed.
The script calculates and displays the following statistics for each test preset:

* basic statistics (min, max, mean, median, standard deviation)
* histogram plot
* plot of subsequent time intervals between events

For some of the statistics, the script also filters out time intervals that deviate from the other records.
These statistics are denoted by a ``no outliers`` suffix in the plot title and output file name.

Apart from displaying the statistics, the scipts also stores them as local files under the :file:`data_stats/` directory.
The results are organized in subdirectories based on the dataset name, time range, and test preset name.

> When running the script from the command line, provide the dataset name (``test1``) and path to the JSON file containing test presets (:file:`stats_nordic_presets/app_event_manager_profiler_tracer.json`).

Reviewer: Please mention that the user can also use regular Profiler sample - event manager is not required.

Author: Added:
`You can use the script to calculate statistics for custom nRF Profiler events or Application Event Manager events, or both.`

If you want to calculate statistics for your own application, you need to create a new test preset file that refers to the nRF Profiler events used by your application.

Test preset JSON file format
----------------------------

The test preset JSON file contains an array of test presets.
Every test preset is identified by a unique name.
The script calculates the time intervals between start and end events specified by each test preset.
The script uses the time intervals as input data for calculating statistics.

An example of an array containing a single test preset looks as follows:

.. code-block:: json

    [
        {
            "name": "Test Name",
            "start_event": {
                "name": "event_name_1",
                "state": "submit"
            },
            "end_event": {
                "name": "event_name_2",
                "state": "processing_end"
            }
        }
    ]

Fields:

* ``name`` - Name of the test preset (used for output files and plot titles)
* ``start_event`` - Configuration for the start event

    * ``name`` - Name of the event

>     * ``state`` - Event state (optional, defaults to "submit").

Reviewer 2: ```suggestion
    * ``state`` - Event state (optional, defaults to "submit")
```

   An application must provide its own implementation.
   See the :ref:`nrf_desktop_nrf_profiler_sync` for an example implementation used by the :ref:`nrf_desktop` application.

.. _nrf_profiler_script_calculating_statistics:

Calculating statistics
======================

The :file:`calc_stats.py` script is used to calculate and visualize nRF Profiler event propagation statistics (statistics for time intervals between nRF Profiler events) for a given dataset.

> Test preset JSON file specifies which events to analyze.

Reviewer 3: I would say, we could use `test preset` or `test configuration` as the json file contains collection of test presets/ test configurations.

> Test preset JSON file specifies which events to analyze.

Author: Changed to `test preset`

> The script calculates and displays the following statistics for every defined test preset:

Reviewer 2: ```suggestion
The script calculates and displays the following statistics for each preset test:
```

* basic statistics (min, max, mean, median, standard deviation),
* histogram plot,
* plot of subsequent time intervals between events.

For some of the statistics, the script also filters out time intervals that are outliers comparing to the other records.

> These statistics are denoted by ``no outliers`` suffix in plot title and output file name.

Reviewer 2: ```suggestion
These statistics are denoted by a ``no outliers`` suffix in the plot title and output file name.
```

> Apart from displaying the statistics, the scipts also stores them as local files under :file:`data_stats/` directory.

Reviewer 2: ```suggestion
Apart from displaying the statistics, the scipts also stores them as local files under the :file:`data_stats/` directory.
```

> The results are organized in subdirectories based on the dataset name, time range, and test preset name.

Reviewer 2: ```suggestion
The results are organized in subdirectories based on the dataset name, time range, and preset test name.
```

> When running the script from the command line, provide the dataset name (``test1``) and path to JSON file containing test presets (``stats_nordic_presets/app_event_manager_profiler_tracer.json``).

Reviewer 2: ```suggestion
When running the script from the command line, provide the dataset name (``test1``) and path to the JSON file containing preset tests (:file:`stats_nordic_presets/app_event_manager_profiler_tracer.json`).
```

.. code-block:: console

    python3 calc_stats.py test1 stats_nordic_presets/app_event_manager_profiler_tracer.json

.. tip::

>    Before calculating statistics using the command example provided above, you need to collect profiling data from a device running the :ref:`app_event_manager_profiling_tracer_sample` sample and store the profiling data as the ``test1`` dataset.

Reviewer 2: ```suggestion
   Before calculating statistics using the example command, you need to collect profiling data from a device running the :ref:`app_event_manager_profiling_tracer_sample` sample and store the data as the ``test1`` dataset.
```

>    Make sure to use the mentioned sample, because the test preset refers to sample-specific nRF Profiler event names.

Reviewer 2: ```suggestion
   Make sure to use the mentioned sample, because the preset test refers to sample-specific nRF Profiler event names.
```

> If you want to calculate statistics for your own application, you need to create a new test preset file which refers to nRF Profiler events used by your application.

Reviewer 2: ```suggestion
If you want to calculate statistics for your own application, you need to create a new preset test file that refers to the nRF Profiler events used by your application.
```

> See the following subsection for details.

Reviewer 2: ```suggestion
```

Test preset JSON file format
----------------------------

The test preset JSON file contains an array of test presets.
Every test preset is identified by a unique name.

> The script calculates time intervals between start and end events specified by every test preset.

Reviewer 2: ```suggestion
The script calculates the time intervals between start and end events specified by each test preset.
```

> Then, the script uses the time intervals as input data for calculating statistics.

Reviewer 2: ```suggestion
The script uses the time intervals as input data for calculating statistics.
```

> An example of the array containing a single test preset looks as follows:

Reviewer 2: ```suggestion
An example of an array containing a single test preset looks as follows:
```

.. code-block:: json

    [
        {
            "name": "Test Name",
            "start_event": {
                "name": "event_name_1",
                "state": "submit"
            },
            "end_event": {
                "name": "event_name_2",
                "state": "processing_end"
            }
        }
    ]

Fields:

* ``name`` - Name of the test preset (used for output files and plot titles)
* ``start_event`` - Configuration for the start event

    * ``name`` - Name of the event
    * ``state`` - Event state (optional, defaults to "submit").

      Possible values:

        * ``submit`` - Event submission
        * ``processing_start`` - Event processing start (only for Application Event Manager events)
        * ``processing_end`` - Event processing end (only for Application Event Manager events)

* ``end_event`` - Configuration for the end event (same format as ``start_event``)

> For examples of the test preset JSON file, check the :file:`stats_nordic_presets/` directory.

Reviewer 2: ```suggestion
For examples of the test preset JSON file, see the :file:`stats_nordic_presets/` directory.
```
